Adding unit tests for publisher output

ycombinator opened the pull request:
## What does this PR do?

Adds unit tests for the publisher output functionality in libbeat.

## Why is it important?

It checks that the publisher output publishes the expected number of events under various scenarios.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] ~I have made corresponding changes to the documentation~
- [ ] ~I have made corresponding change to the default configuration files~
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] ~I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.~

## Related PRs
- In preparation for #17381

elasticmachine commented:
Pinging @elastic/integrations-services (Team:Services)

ycombinator commented on libbeat/publisher/pipeline/output_test.go:
```diff
@@ -0,0 +1,181 @@
+	rand.Seed(time.Now().UnixNano())
+
+	wqu := makeWorkQueue()
+	client := &mockNetworkClient{}
```
The reason I'm only using a `outputs.NetworkClient` here is because that will result in a `netClientWorker` below (instead of a `clientWorker`). And it appears that only in the case of a `netClientWorker` do we cancel the batch if the worker was closed before the batch was published: https://github.com/elastic/beats/blob/8d431691401478bc8d6ed3c80e64b6092e2a332a/libbeat/publisher/pipeline/output.go#L120.

Are we missing a similar check for `clientWorker` inside this loop? https://github.com/elastic/beats/blob/8d431691401478bc8d6ed3c80e64b6092e2a332a/libbeat/publisher/pipeline/output.go#L69-L75

urso replied:
For reloading, I would say we're missing the Cancel for client worker. `clientworker` is only used for console and file output, which I don't think are available via fleet, but better to have the code correct.
Could the reason be that the console and file output do not support a 'Stop/Close' method? Maybe we need to touch those as well?

urso replied:
Just noticed the difference is the `Connect` method. Yeah, clientWorker should also cancel batches in case it is replaced.

urso commented on libbeat/publisher/pipeline/output_test.go:
```diff
@@ -0,0 +1,181 @@
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			rand.Seed(time.Now().UnixNano())
```
Tip: print the seed via `t.Log` and add a CLI flag to configure a static seed. In case randomized testing fails it should be reproducible.

urso commented on libbeat/publisher/pipeline/output_test.go:
```diff
@@ -0,0 +1,181 @@
+					wqu <- batch
+				}()
+			}
+			wg.Wait()
```
for randomized testing `testing/quick` is your friend.

ycombinator replied:
Implemented in 94b133a94.

urso commented on libbeat/publisher/pipeline/output_test.go:
```diff
@@ -0,0 +1,181 @@
+
+			// Make sure that all events have eventually been published
+			c := test.client.(interface{ Published() int })
+			assert.Equal(t, numEvents.Load(), c.Published())
```
better have an assert loop with timeout instead of a hard sleep.

ycombinator replied:
Implemented in 9b1cf7133

urso commented on libbeat/publisher/pipeline/output_test.go:
```diff
@@ -0,0 +1,181 @@
+					numEvents.Add(len(batch.Events()))
+
+					wqu <- batch
+				}()
```
we are moving batches into the work queue only. Why do we need a go-routine for this?
In case the queue/worker blocks the test will timeout (default test timeout in go is 10min).

ycombinator replied:
Yeah, good point. I was trying to simulate as much as possible how the publisher is being used in reality — e.g. multiple FB inputs concurrently trying to publish batches of events. However, the work queue is an unbuffered channel so trying to concurrently send batches to it is pointless 🤦‍♂. I will remove the goroutine.

ycombinator replied:
Fixed in 33af88a42.

urso commented on libbeat/publisher/pipeline/output_test.go:
```diff
@@ -0,0 +1,181 @@
+	assert.Equal(t, numEvents.Load(), client.Published())
+}
+
+type mockClient struct{ published int }
```
we run our tests with `-race`. The way this type is used we should either use an atomic or protect operations via mutex.

ycombinator replied:
Fixed in 33391da48.

ycombinator commented:
@urso I've addressed all your review feedback. It's my first time using `testing/quick` so I hope I'm using it correctly.  Please re-review this PR when you get a chance. Thank you!

ycombinator commented on libbeat/publisher/pipeline/output_test.go:
```diff
@@ -0,0 +1,256 @@
+	}
+}
+
+const publishLatency = 5 * time.Microsecond
```
I'm experimenting with this because the tests are consistently passing without this locally but failing in Travis CI.

ycombinator replied:
Even after trying a few different values, this did not help make the tests pass in Travis CI. So I've removed this constant and it's usages.

ycombinator commented:
@urso I could use some help here with the `TestPublishWithClose` test. This test consistently passes on my laptop and consistently fails on Travis CI. When it fails it fails like so:

```
--- FAIL: TestPublishWithClose (0.74s)
    --- FAIL: TestPublishWithClose/client (0.51s)
        output_test.go:239: reproduce test with `go test ... -seed 1586134537014319105`
        output_test.go:121: 
            	Error Trace:	output_test.go:121
            	            				value.go:460
            	            				value.go:321
            	            				quick.go:290
            	            				output_test.go:93
            	Error:      	"0" is not greater than "0"
            	Test:       	TestPublishWithClose/client
        output_test.go:137: #2: failed on input 0x3212993917d8ebbc
    --- FAIL: TestPublishWithClose/network_client (0.23s)
        output_test.go:239: reproduce test with `go test ... -seed 1586134537525700861`
        output_test.go:121: 
            	Error Trace:	output_test.go:121
            	            				value.go:460
            	            				value.go:321
            	            				quick.go:290
            	            				output_test.go:93
            	Error:      	"0" is not greater than "0"
            	Test:       	TestPublishWithClose/network_client
        output_test.go:137: #1: failed on input 0x1a171ef25f916b87
FAIL
```

Essentially this means that all batches are being consumed by the worker before `Close()` is called on it.

So I tried a couple things (see the last few commits) to try and leave some batches in the queue when `Close()` is called to see if that would make the test pass on Travis CI but it didn't help. Any ideas?

Related: I can only see this test failing on Travis CI right now as the `libbbeat` build is being skipped on Jenkins CI. I rebased this PR on the latest `master` a couple of days ago. Perhaps the Jenkins pipeline dependencies need to be checked?

urso commented on libbeat/publisher/pipeline/output_test.go:
```diff
@@ -0,0 +1,252 @@
+
+				// Close worker before all batches have had time to be published
+				err := worker.Close()
+				require.NoError(t, err)
```
here we might have a race. E.g. how can you tell not all batches have been consumed? Does the initial worker 'block' without ACKing the batch?

ycombinator replied:
The initial worker does not block without ACKing the batch. Indeed, this is probably why I'm seeing different results locally vs. in Travis CI environment.

urso commented on libbeat/publisher/pipeline/output_test.go:
```diff
@@ -0,0 +1,252 @@
+				wg.Wait()
+
+				// Give some time for events to be published
+				timeout := time.Duration(remaining*3) * time.Microsecond
```
thanks to `waitUntilTrue` we can configure a rather big timeout here. E.g. `20s`. This gives systems with low available compute resources a chance to make progress without failing.

ycombinator replied:
Done in 2352fc3.

urso commented:
> So I tried a couple things (see the last few commits) to try and leave some batches in the queue when Close() is called to see if that would make the test pass on Travis CI but it didn't help. Any ideas?

Do you reuse the 'client'? As you switch out the workers, you can have 2 different clients. The first one just 'blocks' (after n batches) and makes no progress, while the second continues processing all batches. The number of CPU/threads might differ on travis, giving you less concurrency and potentially blocking other go-routines for very long.

urso commented:
fyi; the pipeline/stress package has a custom output for testing purposes: https://github.com/elastic/beats/blob/master/libbeat/publisher/pipeline/stress/out.go

ycombinator commented on libbeat/publisher/pipeline/output_test.go:
```diff
@@ -0,0 +1,260 @@
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	// Block publishing
+	if c.publishLimit > 0 && c.published >= c.publishLimit {
+		batch.Retry() // to simulate not acking
+		return nil
+	}
```
@urso WDYT about this logic to emulate blocking publishing?

urso replied:
This emulation is not blocking, but a failing output. The batch is not acked, but this Retrty + return `nil` will signal that the output is ready to consumer another batch.

 A Blocking simulation would require you to wait for some signal (e.g. via a control channel).

ycombinator replied:
At one point I had a sleep in here between the retry and return. The idea then was that the first client would be closed before the sleep finished.

A control channel is better than a sleep. Once the first client is closed I can close the control channel to remove the block. However, the Retry (before waiting to consume from the control channel) will still be needed, otherwise the final publish count doesn't add up to the expected total number of events.

ycombinator replied:
Implemented blocking with control channel in 508b60687.

ycombinator commented on libbeat/publisher/pipeline/output_test.go:
```diff
@@ -0,0 +1,260 @@
+					total := published + client.Published()
+					return numEvents.Load() == total
+				})
+			}, &quick.Config{MaxCount: 50})
```
Keeping the `MaxCount` to 50 seems to keep Travis happy. Before that the build job was being killed because it was taking too long.

urso replied:
Before introducing quick check the count was actually 1 :)

This is some kind of stress test. Unfortunately stress tests don't sit well with travis. We have had bad performance issues with the queue stress tests as well. I think long termI think we should not have stress tests run by travis, but have a separate job running those for even longer. For 'some' simple unit testing a count of 1 might be ok.

ycombinator replied:
True, good point :)

So would you recommend leaving this at 50, lowering it to 1 or maybe somewhere in between? I ask because while 50 is working at the moment I'm worried whether it'll become a source of flakiness. I don't think there's a way to know for sure until Travis runs this several times, though?

urso replied:
well, it's difficult to find the right value. Maybe set it to 25, so we have some more head-room.

urso commented on libbeat/publisher/pipeline/output_test.go:
```diff
@@ -0,0 +1,260 @@
+						numEvents.Add(uint(len(batch.Events())))
+						wqu <- batch
+					}()
+				}
```
instead of creating a go-routine per batch, how about creating a single go-routine that will execute the for loop? This would actually more close to how the workqueue is used.

ycombinator replied:
Sure, I will do that. Just out of curiosity, say there are multiple Filebeat inputs — don't they each get their own goroutine sending to the same publisher work queue?

ycombinator replied:
Implemented in 94f344511.

urso commented on libbeat/publisher/pipeline/output_test.go:
```diff
@@ -0,0 +1,260 @@
+	SeedFlag = flag.Int64("seed", 0, "Randomization seed")
+)
+
+func TestPublish(t *testing.T) {
```
maybe we can rename the test? We don't test publish, but the tests seem to check the behavior of the clientWorkers.

ycombinator replied:
Done in be4bf09cc.

urso commented on libbeat/publisher/pipeline/output_test.go:
```diff
@@ -0,0 +1,260 @@
+type mockClient struct {
+	mu           sync.RWMutex
+	publishLimit uint
+	published    uint
```
maybe we can excert more control by replacing `published` and `publishLimit` with a `func`. The `func` could decide when to block and when to continue. This way the test logic would be more contained within the test. WDYT?

ycombinator replied:
Implemented mockable publish behavior function in 508b60687.

ycombinator commented:
Travis CI is green and Jenkins CI failures are unrelated. Merging.
